### DISCUSSION TITLE
Neighbouring triangles

### DISCUSSION THREAD

**Discussion by micycle1** [gwlucastrig/Tinfour#61] [Q&A] [answered]
Is there a "tinfouronic" way to get the two triangles sharing a given edge? Or neighbouring triangles of a given triangle?

**Comment by gwlucastrig** [comment]
I love your new coinage "tinfouronic"...  That's the first time I've heard that one.  Now I know how Guido von Rossum must have felt :-)

In answer to your question, I note that the Tinfour implementation does not maintain triangle objects internally,  One of my early attempts at the package (call it "tin three") did, but the overhead was too high.  So triangle objects are constructed dynamically, on an as needed basis.

That being said, I think your two recent questions are pointing toward some good enhancements to the API.  I'll have to think a bit more about how to best implement solutions for the requirement you've identified.

For now, I would suggest:

```
          IQuadEdge e;  // an arbitrary edge from the triangulation (vertices A,B)
          SimpleTriangle trig1 = new SimpleTriangle(tin, e, e.getForward(), e.getReverse());
          IQuadEdge d = e.getDual();  // d is the opposite side of e (edge B, A)
          SimpleTriangle trig2 = new SimpleTriangle(tin, d, d.getForward(), d.getReverse());
```

One thing you'll have to watch out for is if the vertices C in the two triangles come back as null references.  This can happen when the edge e is a perimeter edge.

You might find it helpful to give a scan to the wiki page on edge-based techniques at https://github.com/gwlucastrig/Tinfour/wiki/Tutorial-Using-Edge-Based-Techniques-Part-1 
It doesn't answer your questions directly, but it may give some helpful background,

I am thinking that, at the very least, I need to add a constructor to SimpleTriangle 
`
    public SimpleTriangle(IQuadEdge baseEdge);`

**Comment by micycle1** [comment, marked as answer]
I've gone with building up a map of edges-->triangles during `TriangleCollector.visitSimpleTriangles()` and then to find neighbours for a given a triangle, querying the map using the dual of each of its edges.

Curiously, I get ABYSMAL performance when putting *(edge, triangle)* pairs into the map using the IQuadEdge object itself as a key (e.g. `map.put(t.getEdgeA(), t);`). It would seem as though the objects' hashes are colliding but I'm not sure why this is since when I print the hashcodes they seem to be unique. This may be worth an issue in itself.

Using edges as keys:
![image](https://user-images.githubusercontent.com/9304234/108906366-b7cead00-7618-11eb-8ff9-5ff516ca8e79.png)

Using the hashcode as the key (`map.put(t.getEdgeB().hashCode(), t);`) results in expected `put()` performance:

![image](https://user-images.githubusercontent.com/9304234/108907025-7c80ae00-7619-11eb-8fee-fa9551cd0cd4.png)

In which case getting a triangle's neighbours now looks like this:

```
SimpleTriangle live;
final SimpleTriangle n1 = map.get(live.getEdgeA().getDual().hashCode());
final SimpleTriangle n2 = map.get(live.getEdgeB().getDual().hashCode());
final SimpleTriangle n3 = map.get(live.getEdgeC().getDual().hashCode());
```

**Reply by micycle1** [reply]
Found the cause of the hash collisions: https://github.com/gwlucastrig/Tinfour/issues/62.

**Reply by gwlucastrig** [reply]
Thanks.  I use Netbeans as my IDE and I just asked it to generate a has code.  And it did the same foolish thing.  So now I know how the bad hash got in there.   Serves me right for trusting automatically generated code. What did you end up using?

The index is essentially the unique ID of the edge.  So I suppose a good hash need only use the index and not worry about the other fields.  Also, for QuadEdgePartner, you should definitely use the getIndex() method rather than accessing the index field directly because the index field is overloaded (metaphorically speaking) in order to conserve memory.
